### ISSUE TITLE
Block User

### ISSUE THREAD

**Issue by Lokke** [humhub/humhub#564]
there should be a option to block users

**Comment by andystrobel** [comment]
Which kind of blocking you mean, exactly? You can already verify users at registration and disable users at Administration -> users.

**Comment by alirz23** [comment]
I think he means black list users so the can't post or update stuff

**Comment by WASasquatch** [comment]
He means that a registered use shuold be able to block any user. Next to "Follow/Unfollow" there could be a small "Block" button (or hidden in a dropdown menu of other useful links for the future) that allows a user to block another user. IE, once this happens the users can no longer interact, see each others profile, or posts.

**Comment by thedude83** [comment]
I'd have to agree that this is a vital feature.

**Comment by enginhu** [comment]
block user is necessary feature.

**Comment by nikolaaa4** [comment]
Any update on this?

**Comment by lipsmega** [comment]
updates?

**Comment by michaelh99** [comment]
I'm guessing this is considered an undesirable feature and that moderation is the preferred method for handling conflicts?

**Comment by WASasquatch** [comment]
I'm rather shocked that this, of all features, hasn't been introduced yet. This is somewhat why I left HumHub years ago. Development seemed to be focusing on the wrong areas trying to get "purchasable" features, rather than focusing on basic neceassary features to constitute a purchase to begin with.

**Comment by Ace2213** [comment]
Blocking would allow any user to block another user, to avoid contacting them and seeing their posts and profile. This is an essential privacy feature in 2019 to give users control over who interacts with them and sees them.

Particularly in the case of stalkers, abusers etc without having to delete your account or set everything to private and be paranoid.

From an admin's point of view, it provides an alternative method of handling conflict without needing to delete someone's account every time there's a disagreement.

It's probably not very easy to implement, since you would have to add an entire blocking system that provides exceptions to every existing rule that governs streams, user picker, etc as well as a rule to make any page related to someone that has blocked you return a 404 error, to also not make it clear that the person has been blocked as that can sometimes make the person even more abusive.

**Comment by buddh4** [comment]
Yeah, almost only [purchasable features](https://www.humhub.org/de/marketplace), those greedy developers... As @Ace2213 mentioned we need a bullet proof concept for this first, since it will probably affect almost all modules.

**Comment by DiscordiaChaos** [comment]
I was wondering about this, a user was asking me the other day how they could block someone that keeps sending unwanted messages and friend requests. The unwanted messages are not enough to remove the messager, so it would be preferred that users can block each other if they need to in order to prevent a user from just leaving or using the site less to avoid another user they dislike. It is such a common feature in social networks and forums I was a bit surprised it doesn't exist yet.

**Comment by ArchBlood** [comment]
> I was wondering about this, a user was asking me the other day how they could block someone that keeps sending unwanted messages and friend requests. The unwanted messages are not enough to remove the messager, so it would be preferred that users can block each other if they need to in order to prevent a user from just leaving or using the site less to avoid another user they dislike. It is such a common feature in social networks and forums I was a bit surprised it doesn't exist yet.

At the moment the only thing I can suggest is to have them disable messages by the permissions
![Screenshot_1](https://user-images.githubusercontent.com/35392110/109433868-896a1c80-79e0-11eb-9d52-ba84c484cdb0.png)

**Comment by DiscordiaChaos** [comment]
Thank you @ArchBlood , I have notified and instructed the user on how to do this. :)

**Comment by WASasquatch** [comment]
Seems weird to me no one has done this yet in development. You just need to
create a new column for the users and store an array of IDs or something
like that. Make sure to check that the user being blocked doesn't have
elevated permissions over the blockee. Then in content parsing you check if
a user can see the content being rendered. If not, simply skip it. For
posts, and comments, and member lists. Add a separate class for checking
for bans etc so plugins can implement their own checks, and bam.

**Comment by luke-** [comment]
**Concept**

- Add  User Picker with `Blocked users` in `Account -> Settings`
- `Mail Modul` should use these users  
- Don't show any entries in Stream & Activites of blocked users. Inside module views e.g. Files module, regulary show contents of blocked users 
- Don't allow access to own profile if visitor is blocked
- Don't allow Mentioning by blocked users
- Comments of a blocked users: Only show line "Comment of blocked user [Show]"
- No notifications of any blocked users

**Comment by yurabakhtin** [comment]
@luke- 

> - Add User Picker with `Blocked users` in `Account -> Settings`

I have started this in PR https://github.com/humhub/humhub/pull/5309:

<img width="1147" alt="blocked_users" src="https://user-images.githubusercontent.com/6995524/134693552-1a388681-ac9e-4245-8b9e-d27cfa8b5ee2.png">

**Comment by yurabakhtin** [comment]
@luke- 

> - Don't allow access to own profile if visitor is blocked

Commit https://github.com/humhub/humhub/pull/5309/commits/24ae53ff4d483943953e0486a3d80264cae035bf.

**Comment by luke-** [comment]
@yurabakhtin Looks good so far. With the block, a new DB query is necessary for each ContentContainer (Space/Profile) page call (check if user is blocked).

What do you think about additionally serializing these user ids as contentcontainer setting? Settings are very performant, cached, read as bunch and have to be loaded anyway.

**Comment by yurabakhtin** [comment]
@luke- 

> What do you think about additionally serializing these user ids as contentcontainer setting? Settings are very performant, cached, read as bunch and have to be loaded anyway.

I created new table `contentcontainer_blocked_users` with columns `contentcontainer_id` and `user_id`.
If we will store `user_id`s as serialized string per container then it will not impossible to do quick filter user in single SQL query, I mean sql `JOIN` if it will be required in future, i.e. if you think it is normal to keep a code like this `->andWhere(['NOT IN', 'user.id', $container->settings->get('blockedUserIds')])` instead of `->leftJoin('contentcontainer_blocked_users', 'contentcontainer_blocked_users.user_id = user.id')->where('contentcontainer_blocked_users.user_id IS NULL')` then we can store all user ids in single setting string and if you are sure the container settings always loaded anyway.

**Comment by luke-** [comment]
@yurabakhtin The current code with an table `contentcontainer_blocked_users` is great!  I would fill the settings **additionally**. e.g. at `contentcontainer_blocked_users::afterSave()` update the settings for the respective ContentContainer. **As additional cache.ÜÜ

**Comment by luke-** [comment]
Specifically, I am only concerned with this line: https://github.com/humhub/humhub/blob/d69de26d5df9617afedb66301fa5ba02f97af667/protected/humhub/modules/content/components/ContentContainerController.php#L92

This would create an additional DB query on every ContentContainer request (also partials). If this can somehow **cached** as `ContentContainerSetting` , we could hopefully avoid it.

**Comment by yurabakhtin** [comment]
@luke- 

> I would fill the settings **additionally**.

Ok, got it now, I will implement this.

**Comment by luke-** [comment]
@yurabakhtin perfect, thanks! :-)

**Comment by yurabakhtin** [comment]
@luke- 

> - Don't show any entries in Stream & Activites of blocked users.

Commit https://github.com/humhub/humhub/pull/5309/commits/791b0c869a1689c30a09329603403bf275cc88a5.

> Inside module views e.g. Files module, regulary show contents of blocked users

Sorry, not clear what you mean here especially words "regulary show", maybe you wanted to say "do not show"?

**Comment by luke-** [comment]
@yurabakhtin That's correct. We could also say that only stream entries of blocked users should **not** be displayed.
But in e.g. the Wiki module itself, the Wiki pages or in the CFiles module the files of blocked users should be displayed. 
Is this clearer? :-)

**Comment by yurabakhtin** [comment]
@luke- 

> But in e.g. the Wiki module itself, the Wiki pages or in the CFiles module the files of blocked users should be displayed.
Is this clearer? :-)

Yes, clearer, thanks. I filtered only the stream query, so all other modules are not filtered inside by blocked users yet.

**Comment by luke-** [comment]
Ok, so it should be :-)

**Comment by yurabakhtin** [comment]
@luke- 

> - Comments of a blocked users: Only show line "Comment of blocked user [Show]"

Commit https://github.com/humhub/humhub/pull/5309/commits/d5eebb18027856c9aee8c40197959d63d1c771a3:

<img width="667" alt="hide_blocked_comment" src="https://user-images.githubusercontent.com/6995524/135245016-2aa1b0ae-7722-4f04-8c5c-a002cb017860.png">

**Comment by luke-** [comment]
@orangiene Can you take a look here? Is the button ok? Or should we display a simple link here? Or the display name of the blocked user?

**Comment by orangiene** [comment]
@yurabakhtin 
I talked to luke, so let's try this:
can you please use a dummy image as a profile-image in black/white, next to it the text "Comment of blocked user" und an usual link in blue "show comment"?
is this clear or should I draw a quick sketch?

**Comment by yurabakhtin** [comment]
@orangiene yes, all clear, thanks.

**Comment by yurabakhtin** [comment]
@luke- @orangiene Commit https://github.com/humhub/humhub/pull/5309/commits/a81cb4ef4735ae8e8847d9320e48a9967b44f8e5:

<img width="662" alt="comment_view_changed" src="https://user-images.githubusercontent.com/6995524/135255359-fa1486bf-97cf-4080-b828-e99393f97496.png">

**Comment by luke-** [comment]
@yurabakhtin @orangiene Looks good! Thank you

@Semir1212 Can you please check/approve the wording?   Maybe link text e.g. "View comment anyway".

**Comment by luke-** [comment]
@yurabakhtin One missed requirement: The admin should be able to control the "User Blocking via a setting (Administration -> User -> Settings). Maybe just a checkbox that deletes all user blocks and hides the user picker when disabled.

And tests for the individual areas would also be good. :-)

**Comment by Semir1212** [comment]
@luke- Looks good, but I would prefer just 'Show"' instead of 'Show comment' to avoid having 2x 'comment' within one line.

**Comment by yurabakhtin** [comment]
@Semir1212 @luke- 

> Looks good, but I would prefer just 'Show"' instead of 'Show comment' to avoid having 2x 'comment' within one line.

Commit https://github.com/humhub/humhub/pull/5309/commits/884003323d569173f1d8665bf20afde5b4e9214e.

**Comment by cargsource** [comment]
Hello, I am new to this and was wondering how to add this to my website? I would very much like a blocking mechanism for my users.

**Comment by ArchBlood** [comment]
> Hello, I am new to this and was wondering how to add this to my website? I would very much like a blocking mechanism for my users.

Best to wait for the next major release as I'm sure that many changes will be done before it's fully implemented.

**Comment by yurabakhtin** [comment]
@luke- 

> One missed requirement: The admin should be able to control the "User Blocking via a setting (Administration -> User -> Settings). Maybe just a checkbox that deletes all user blocks and hides the user picker when disabled.

Commit https://github.com/humhub/humhub/pull/5309/commits/cff6a1bbf3913a86aee8ee651d8021fbfc1db923:

<img width="1148" alt="allow_block_users" src="https://user-images.githubusercontent.com/6995524/136010313-49929b15-ef57-41a2-b4c6-6b55622406d2.png">

Why do you want to delete all users blocks? The checkbox may be unchecked by mistake, so it will be impossible to revert the deleted data, I think it is enough to disable the feature by condition as I implemented in the commit.

**Comment by luke-** [comment]
@yurabakhtin Ok, I thought it might be easier and better in terms of performance. But querying the setting should also be performant.

**Comment by luke-** [comment]
@Semir1212 `Allow to block users`  Any idea how to label this checkbox?

**Comment by Semir1212** [comment]
@luke- Allow users to block each other

**Comment by luke-** [comment]
@yurabakhtin Can you please rename the checkbox label?

**Comment by yurabakhtin** [comment]
@luke- 

> Can you please rename the checkbox label?

Done - https://github.com/humhub/humhub/pull/5309/commits/f29e25830ae27dbe79cd1b7311d999ef8d095b96#diff-37a7685bbfd37cce1f083ff91f4701c54c88f9455e25f5443dfaa065268f340bR81.

> - No notifications of any blocked users

Commit https://github.com/humhub/humhub/pull/5309/commits/f29e25830ae27dbe79cd1b7311d999ef8d095b96.

**Comment by yurabakhtin** [comment]
@luke- 

> - Don't allow Mentioning by blocked users

Commit https://github.com/humhub/humhub/pull/5309/commits/740ecfcc7afff506f6b7f5ab9d492c420b8d4211.

**Comment by yurabakhtin** [comment]
From this list https://github.com/humhub/humhub/issues/564#issuecomment-916121141 only the point is not completed yet:

- `Mail Module` should use these users

**Comment by luke-** [comment]
@yurabakhtin Ok, that means I can already review the core part? Of course, a few more tests would be good. :-)

**Comment by yurabakhtin** [comment]
@luke- Yes, please review that because all points related to core were done except of tests.

**Comment by luke-** [comment]
@yurabakhtin Thanks, nice work! I've added my review.

**Comment by yurabakhtin** [comment]
@luke- 

> - `Mail Module` should use these users

I have started this in PR https://github.com/humhub/mail/pull/238, what I have done:

- Exclude the blocked users from the recipients picker on creating new conversation
- Make profile pictures grayscale for all blocked users:

<img width="955" alt="messaging_blocked_users" src="https://user-images.githubusercontent.com/6995524/136198540-957e991c-aafe-40a1-a520-7973cae1b6d1.png">

- Don't allow to reply if at least one recipient is blocked for the current User. Probably we should hide the reply form completely with message like "This conversation is closed for you because you are blocked for the users First User, Second User, Third User.", do you agree?

**Comment by luke-** [comment]
@yurabakhtin Yes, sounds good!

**Comment by luke-** [comment]
@Semir1212 Text please :-)

**Comment by yurabakhtin** [comment]
@Semir1212 The text is updated in commit https://github.com/humhub/mail/pull/238/commits/b7f0790feda535f32ad064b76e5de5d9f6d1f9dd.

**Comment by michaelh99** [comment]
First, apologies for replying on an old issue.  I didn't see a better place to ask and the functionality I wanted to ask about is similar to blocking but google has been no help.

Is there any way to have a middle ground between seeing everything by a user and completely blocking them?

For example, I'd like to be able to "mute" a user so that notifications of their posts never show in my stream or daily digests in my inbox but otherwise not block them.  Another way of thinking about it would be "ghosting" the other user.  I never see their contributions but they have no way to know that.

**Comment by ArchBlood** [comment]
> First, apologies for replying on an old issue.  I didn't see a better place to ask and the functionality I wanted to ask about is similar to blocking but google has been no help.
> 
> Is there any way to have a middle ground between seeing everything by a user and completely blocking them?
> 
> For example, I'd like to be able to "mute" a user so that notifications of their posts never show in my stream or daily digests in my inbox but otherwise not block them.  Another way of thinking about it would be "ghosting" the other user.  I never see their contributions but they have no way to know that.

See #4188

**Comment by michaelh99** [comment]
@ArchBlood  thank you, that's very similar.  I'll start a new topic because though what I'm asking for is close, there are important differences

**Comment by ArchBlood** [comment]
> @ArchBlood  thank you, that's very similar.  I may start a new topic because though what I'm asking for is close, there are important differences.
> 
> One further question about this feature, I've enabled user blocking in the admin settings but as a user I can't find any way to block someone.  I don't see anything in the user profile page, user list or user management in the admin pages.  What am I missing?

Check user account settings
